authenticate/handlers.go

```diff
@@ -100,71 +101,84 @@ func (a *Authenticate) refresh(w http.ResponseWriter, r *http.Request, s *sessio
 // RobotsTxt handles the /robots.txt route.
 func (a *Authenticate) RobotsTxt(w http.ResponseWriter, r *http.Request) {
 	w.Header().Set("Content-Type", "text/plain; charset=utf-8")
-	w.Header().Set("X-Content-Type-Options", "nosniff")
 	w.WriteHeader(http.StatusOK)
 	fmt.Fprintf(w, "User-agent: *\nDisallow: /")
 }
 
 // SignIn handles to authenticating a user.
 func (a *Authenticate) SignIn(w http.ResponseWriter, r *http.Request) {
-	// grab and parse our redirect_uri
-	redirectURL, err := urlutil.ParseAndValidateURL(r.FormValue("redirect_uri"))
+	redirectURL, err := urlutil.ParseAndValidateURL(r.FormValue(urlutil.QueryRedirectURI))
 	if err != nil {
 		httputil.ErrorResponse(w, r, httputil.Error("malformed redirect_uri", http.StatusBadRequest, err))
 		return
 	}
-	// create a clone of the redirect URI, unless this is a programmatic request
-	// in which case we will redirect back to proxy's callback endpoint
-	callbackURL, _ := urlutil.DeepCopy(redirectURL)
 
-	q := redirectURL.Query()
+	jwtAudience := []string{a.RedirectURL.Hostname(), redirectURL.Hostname()}
 
-	if q.Get("pomerium_programmatic_destination_url") != "" {
-		callbackURL, err = urlutil.ParseAndValidateURL(q.Get("pomerium_programmatic_destination_url"))
+	var callbackURL *url.URL
+	// if the callback is explicitly set, set it and add an additional audience
+	if callbackStr := r.FormValue(urlutil.QueryCallbackURI); callbackStr != "" {
+		callbackURL, err = urlutil.ParseAndValidateURL(callbackStr)
 		if err != nil {
-			httputil.ErrorResponse(w, r, httputil.Error("", http.StatusBadRequest, err))
+			httputil.ErrorResponse(w, r, httputil.Error(err.Error(), http.StatusBadRequest, err))
 			return
 		}
+		jwtAudience = append(jwtAudience, callbackURL.Hostname())
+	} else {
+		// otherwise, assume callback is the same host as redirect
+		callbackURL, _ = urlutil.DeepCopy(redirectURL)
```

Comment:
Is that an error we should be checking?

Reply:
https://github.com/pomerium/pomerium/blob/aac36a8bdb0b1a0639a5fda204d6b544281ef496/authenticate/handlers.go#L110

I think safe in this context but I'll add a comment.